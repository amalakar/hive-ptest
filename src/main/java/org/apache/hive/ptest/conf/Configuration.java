/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.hive.ptest.conf;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;
import java.util.Set;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Joiner;
import com.google.common.base.Preconditions;
import com.google.common.base.Splitter;
import com.google.common.base.Strings;
import com.google.common.collect.Iterables;
import com.google.common.collect.Maps;
import com.google.common.collect.Sets;

public class Configuration {
  
  private static final Logger LOG = LoggerFactory
      .getLogger(Configuration.class);


  private final Set<Host> hosts;
  private final String privateKey;
  private final String workingDirectory;
  private final Context context;
  private String antArgs;
  private String antEnvOpts;
  private String repository;
  private String repositoryName;
  private String patch;
  private String javaHome;
  private String branch;
  
  @VisibleForTesting
  public Configuration(Context context) {
    this.context = context;
    hosts = Sets.newHashSet();
    for(String alias : Splitter.on(" ").omitEmptyStrings().split(context.getString("hosts", ""))) {
      Context hostContext = new Context(context.getSubProperties(
          Joiner.on(".").join("host", alias, "")));
      LOG.info("Processing host {}: {}", alias, hostContext.getParameters().toString());
      hosts.add(new Host(hostContext.getString("host"), hostContext.getString("user"),
          Iterables.toArray(Splitter.on(",").trimResults().split(hostContext.getString("localDirs")), String.class),
          hostContext.getInteger("threads")));
    }
    Preconditions.checkState(hosts.size() > 0, "no hosts specified");
    repository =  Preconditions.checkNotNull(context.getString("repository"), "repository").trim();
    repositoryName =  Preconditions.checkNotNull(context.getString("repositoryName"), "repositoryName").trim();
    privateKey =  Preconditions.checkNotNull(context.getString("privateKey"), "privateKey").trim();
    branch =  Preconditions.checkNotNull(context.getString("branch"), "branch").trim();
    workingDirectory =  Preconditions.checkNotNull(context.getString("workingDirectory"), 
        "workingDirectory").trim();
    antArgs =  Preconditions.checkNotNull(context.getString("antArgs"), "antArgs").trim();
    antEnvOpts =  context.getString("antEnvOpts", "").trim();
    javaHome =  context.getString("javaHome", "").trim();    
  }
  public Set<Host> getHosts() {
    return hosts;
  }
  public Context getContext() {
    return context;
  }
  public String getRepositoryName() {
    return repositoryName;
  }
  public String getRepository() {
    return repository;
  }

  public String getBranch() {
    return branch;
  }

  public String getPrivateKey() {
    return privateKey;
  }
  public String getWorkingDirectory() {
    return workingDirectory;
  }
  public String getAntArgs() {
    return antArgs;
  }
  public String getJavaHome() {
    return javaHome;
  }
  public String getPatch() {
    return patch;
  }
  public void setPatch(String patch) {
    this.patch = Strings.nullToEmpty(patch);
  }
  public void setRepository(String repository) {
    this.repository = Strings.nullToEmpty(repository);
  }
  public void setRepositoryName(String repositoryName) {
    this.repositoryName = Strings.nullToEmpty(repositoryName);
  }
  public void setBranch(String branch) {
    this.branch = Strings.nullToEmpty(branch);
  }
  public void setJavaHome(String javaHome) {
    this.javaHome = Strings.nullToEmpty(javaHome);
  }
  public void setAntArgs(String antArgs) {
    this.antArgs = Strings.nullToEmpty(antArgs);
  }  
  public String getAntEnvOpts() {
    return antEnvOpts;
  }
  public void setAntEnvOpts(String antEnvOpts) {
    this.antEnvOpts = Strings.nullToEmpty(antEnvOpts);
  }
  @Override
  public String toString() {
    return "Configuration [privateKey=" + privateKey + ", workingDirectory="
        + workingDirectory + ", context=" + context + ", antArgs=" + antArgs
        + ", antEnvOpts=" + antEnvOpts + ", repository=" + repository
        + ", repositoryName=" + repositoryName + ", patch=" + patch
        + ", javaHome=" + javaHome + ", branch=" + branch + "]";
  }
  public static Configuration fromInputStream(InputStream inputStream)
      throws IOException {
    Properties properties = new Properties();
    properties.load(inputStream);
    Context context = new Context(Maps.fromProperties(properties));
    return new Configuration(context);
  }
  public static Configuration fromFile(String file) throws IOException {
    return fromInputStream(new FileInputStream(file));
  }
}
