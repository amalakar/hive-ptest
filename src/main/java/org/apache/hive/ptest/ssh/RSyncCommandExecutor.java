/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.hive.ptest.ssh;

import java.io.IOException;

import org.apache.hive.ptest.LocalCommand;
import org.apache.hive.ptest.LocalCommand.CollectPolicy;


public class RSyncCommandExecutor {

  public void execute(RSyncCommand command) {
    CollectPolicy collector = new CollectPolicy();
    try {
      LocalCommand cmd;
      if(command.getType() == RSyncCommand.Type.TO_LOCAL) {
        throw new UnsupportedOperationException(String.valueOf(command.getType()));
      } else if(command.getType() == RSyncCommand.Type.FROM_LOCAL) {
        cmd = new LocalCommand(collector,
            String.format("rsync -qavPe \"ssh -i %s\" --delete --delete-during --force %s %s@%s:%s", 
                command.getPrivateKey(), command.getLocalFile(), command.getUser(), command.getHost(),
                command.getRemoteFile()));
      } else {
        throw new UnsupportedOperationException(String.valueOf(command.getType()));
      }
      command.setExitCode(cmd.getExitCode());
    } catch (IOException e) {
      command.setException(e);
    } catch (InterruptedException e) {
      command.setException(e);
    } finally {
      command.setOutput(collector.getOutput());
    }
  }
}
