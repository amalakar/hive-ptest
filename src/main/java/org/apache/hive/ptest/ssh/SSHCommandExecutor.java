/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.hive.ptest.ssh;

import org.apache.hive.ptest.LocalCommand;
import org.apache.hive.ptest.LocalCommand.CollectPolicy;

public class SSHCommandExecutor {

  public void execute(SSHCommand command) {
    CollectPolicy collector = new CollectPolicy();
    try {
      LocalCommand cmd = new LocalCommand(collector, 
          String.format("ssh -i %s -l %s %s '%s'", command.getPrivateKey(), 
              command.getUser(), command.getHost(), command.getCommand()));
      command.setExitCode(cmd.getExitCode());
    } catch (Exception e) {
      command.setException(e);
    } finally {
      command.setOutput(collector.getOutput());
    }
  }
}
