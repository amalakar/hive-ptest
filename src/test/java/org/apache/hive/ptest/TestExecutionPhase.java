/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.hive.ptest;

import static org.mockito.Mockito.when;

import java.io.File;
import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.Set;

import org.apache.commons.io.FileUtils;
import org.apache.hive.ptest.conf.QFileTestBatch;
import org.apache.hive.ptest.conf.TestBatch;
import org.apache.hive.ptest.conf.UnitTestBatch;
import org.apache.hive.ptest.ssh.NonZeroExitCodeException;
import org.apache.hive.ptest.ssh.SSHCommand;
import org.approvaltests.Approvals;
import org.approvaltests.reporters.JunitReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import com.google.common.base.Charsets;
import com.google.common.base.Suppliers;
import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import com.google.common.io.Files;

@UseReporter(JunitReporter.class)
public class TestExecutionPhase extends AbstractTestPhase {
  private static final String DRIVER = "driver";
  private static final String QFILENAME = "sometest";
  private ExecutionPhase phase;
  private File baseDir;
  private File scriptDir;
  private File logDir;
  private File succeededLogDir;
  private File failedLogDir;
  private File testDir;
  private List<String> messages;
  private Set<String> failedTests;
  private List<TestBatch> testBatches;
  private TestBatch testBatch;

  
  @Before
  public void setup() throws Exception {
    super.setup();
    baseDir = createBaseDir(this.getClass().getSimpleName());
    scriptDir = Dirs.create(new File(baseDir, "scripts"));
    logDir = Dirs.create(new File(baseDir, "logs"));
    succeededLogDir = Dirs.create(new File(logDir, "succeeded"));
    failedLogDir = Dirs.create(new File(logDir, "failed"));
    scriptDir = Dirs.create(new File(baseDir, "scripts"));
    messages = Lists.newArrayList();
    failedTests = Sets.newHashSet();
  }
  private ExecutionPhase getPhase() throws IOException {
    phase = new ExecutionPhase(drones, executor, sshCommandExecutor, 
        scpCommandExecutor, rsyncCommandExecutor, localCommandFactory, templateDefaults, baseDir,
        scriptDir, succeededLogDir, failedLogDir, messages, Suppliers.ofInstance(testBatches), 
        failedTests);
    return phase;
  }
  private void setupQFile(boolean isParallel) throws Exception {
    testDir = Dirs.create( new File(baseDir, "test"));
    Assert.assertTrue(new File(testDir, QFILENAME).createNewFile());
    testBatch = new QFileTestBatch(DRIVER, "qfile", Sets.newHashSet(QFILENAME), isParallel);
    testBatches = Collections.singletonList(testBatch);
  }
  private void setupUnitTest() throws Exception {
    testBatch = new UnitTestBatch(DRIVER, false);
    testBatches = Collections.singletonList(testBatch);
  }
  @After
  public void teardown() {
    FileUtils.deleteQuietly(baseDir);
  }
  @Test
  public void testPassingQFileTest() throws Throwable {
    setupQFile(true);
    getPhase().execute();
    Approvals.verify(getExecutedCommands());
    Assert.assertEquals(Sets.newHashSet(), failedTests);
    Assert.assertEquals(1, messages.size());
  }
  @Test
  public void testFailingQFile() throws Throwable {
    setupQFile(true);
    sshCommandExecutor = new MockSSHCommandExecutor() {
      public void execute(SSHCommand command) {
        commands.add(command.getCommand());
        command.setExitCode(1);
        command.setOutput("");
      }
    };
    String junitOutput = Phase.readResource("TEST-SomeTest-failure.xml");
    File driverFailureDir = Dirs.create(new File(failedLogDir, testBatch.getName()));
    File junitOutputFile = new File(driverFailureDir, "TEST-SomeTest-failure.xml");
    Files.write(junitOutput.getBytes(Charsets.UTF_8), junitOutputFile);
    try {
      getPhase().execute();
      Assert.fail("Expected TestsFailedException");
    } catch (TestsFailedException e) {
     Assert.assertEquals("1 test batches failed", e.getMessage());
    }
    Approvals.verify(getExecutedCommands());
    Assert.assertEquals(Sets.newHashSet("SomeTest." + QFILENAME), failedTests);
    Assert.assertEquals(2, messages.size());
  }
  @Test
  public void testPassingUnitTest() throws Throwable {
    setupUnitTest();
    getPhase().execute();
    Approvals.verify(getExecutedCommands());
    Assert.assertEquals(Sets.newHashSet(), failedTests);
    Assert.assertEquals(1, messages.size());
  }
  @Test
  public void testFailingUnitTest() throws Throwable {
    setupUnitTest();
    when(localCommand.getExitCode()).thenReturn(1, 0);
    try {
      getPhase().execute();
      Assert.fail();
    } catch (TestsFailedException e) {
      Assert.assertEquals("1 test batches failed", e.getMessage());
    }
    Approvals.verify(getExecutedCommands());
    Assert.assertEquals(Sets.newHashSet(), failedTests);
    Assert.assertEquals(messages.toString(), 2, messages.size());
  }
  @Test(expected = NonZeroExitCodeException.class)
  public void testFailedCopy() throws Throwable {
    setupUnitTest();
    when(localCommand.getExitCode()).thenReturn(1);
    getPhase().execute();
  }
}