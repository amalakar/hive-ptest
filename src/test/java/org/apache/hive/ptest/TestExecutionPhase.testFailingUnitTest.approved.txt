bash /tmp/hive-ptest-units/TestExecutionPhase/isolated/hiveptest-driver.sh
cp -R /tmp/hive-ptest-units/TestExecutionPhase/isolated/logs/* /tmp/hive-ptest-units/TestExecutionPhase/logs/failed/driver